[WebGPU] Implement GPUAdapter.requestAdapterInfo

mwyrzykowski opened the pull request:
#### 1e0ff03a5f0552323e52f26820db1839dac9c864
<pre>
[WebGPU] Implement GPUAdapter.requestAdapterInfo
<a href="https://bugs.webkit.org/show_bug.cgi?id=251377">https://bugs.webkit.org/show_bug.cgi?id=251377</a>
&lt;radar://104829439&gt;

Reviewed by Myles C. Maxfield.

Minimally implement requestAdapterInfo, so <a href="https://mlc.ai/web-llm/">https://mlc.ai/web-llm/</a>
does not report an error about output.adapterInfo.description being
undefined.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUAdapter.cpp:
(WebCore::GPUAdapter::requestAdapterInfo):
* Source/WebCore/Modules/WebGPU/GPUAdapter.h:
* Source/WebCore/Modules/WebGPU/GPUAdapter.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.h: Added.
(WebCore::GPUAdapterInfo::create):
(WebCore::GPUAdapterInfo::vendor const):
(WebCore::GPUAdapterInfo::architecture const):
(WebCore::GPUAdapterInfo::device const):
(WebCore::GPUAdapterInfo::description const):
(WebCore::GPUAdapterInfo::GPUAdapterInfo):
* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.idl: Added.
Add new files.

Canonical link: <a href="https://commits.webkit.org/263164@main">https://commits.webkit.org/263164@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a2b6895c1885a71a78065d1bb9bfa861f2d84626

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3919 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4027 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5263 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4089 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3805 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3998 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3924 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3674 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3877 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4076 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3450 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5097 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1580 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3425 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/4709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3399 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3485 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4865 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3882 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/3141 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3407 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3425 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/929 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3448 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3684 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

litherum commented on Source/WebCore/Modules/WebGPU/GPUAdapter.cpp:
```diff
@@ -72,8 +73,7 @@ void GPUAdapter::requestDevice(ScriptExecutionContext&, const std::optional<GPUD
 {
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=251377 - [WebGPU] Implement GPUAdapter.requestAdapterInfo
-    promise.resolve(nullptr);
+    promise.resolve(GPUAdapterInfo::create());
```
I'm not sure this is sufficient - we will probably want this to be more complicated in the future.

I think we should have device creation automatically report this info from the GPU Process to the web process - like how limits and features are already handled

mwyrzykowski replied:
WebGPUAdapter already has an adapter name, we could likely use this directly to populate at least one of the GPUAdapterInfo fields, I will do that for now. We can expand upon it as needed in the future

mwyrzykowski committed:
[WebGPU] Implement GPUAdapter.requestAdapterInfo
https://bugs.webkit.org/show_bug.cgi?id=251377
<radar://104829439>

Reviewed by Myles C. Maxfield.

Minimally implement requestAdapterInfo, so https://mlc.ai/web-llm/
does not report an error about output.adapterInfo.description being
undefined.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUAdapter.cpp:
(WebCore::GPUAdapter::requestAdapterInfo):
* Source/WebCore/Modules/WebGPU/GPUAdapter.h:
* Source/WebCore/Modules/WebGPU/GPUAdapter.idl:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:

* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.h: Added.
(WebCore::GPUAdapterInfo::create):
(WebCore::GPUAdapterInfo::vendor const):
(WebCore::GPUAdapterInfo::architecture const):
(WebCore::GPUAdapterInfo::device const):
(WebCore::GPUAdapterInfo::description const):
(WebCore::GPUAdapterInfo::GPUAdapterInfo):
* Source/WebCore/Modules/WebGPU/GPUAdapterInfo.idl: Added.
Add new files.

Canonical link: https://commits.webkit.org/263164@main
